### PR TITLE
Publicize required methods to allow for custom implementations of resource detectors.

### PR DESCRIPTION
**Pull request by Yun-Ting**
## Changes
Addressed https://github.com/open-telemetry/opentelemetry-dotnet/pull/2897#issuecomment-1040385720
https://github.com/open-telemetry/opentelemetry-dotnet/pull/2897 alone itself was not sufficient for the users to implement their own resource detector. And this PR bridge this gap.

For significant contributions please make sure you have completed the following items:

* [x] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [x] Changes in public API reviewed

### CODE REVIEW

File: docs/trace/extending-the-sdk/MyResourceDetector.cs
Comment (reyang) [inline docs/trace/extending-the-sdk/MyResourceDetector.cs:21]: How is this being used in this example?

File: docs/trace/extending-the-sdk/MyResourceDetector.cs
Comment (reyang) [inline docs/trace/extending-the-sdk/MyResourceDetector.cs:24]: Might need to update the docs https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/docs/trace/extending-the-sdk#extending-the-opentelemetry-net-sdk

File: docs/trace/extending-the-sdk/MyResourceDetector.cs
Comment (reyang) [inline docs/trace/extending-the-sdk/MyResourceDetector.cs:25]: The message seems a bit weird.

File: docs/trace/extending-the-sdk/MyResourceDetector.cs
Comment (reyang) [inline docs/trace/extending-the-sdk/MyResourceDetector.cs:31]: nit: might be simplified to `return new Resource(attributes)`.

File: src/OpenTelemetry/Resources/Resource.cs
Comment (cijothomas) [inline src/OpenTelemetry/Resources/Resource.cs:38]: please keep this internal as before.
We already exposed IResourceDetector, and we are missing a public method to add detectors : https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Resources/ResourceBuilder.cs#L89-L91 --> this is the thing to public ize to address the issue.

**Review by cijothomas — CHANGES_REQUESTED** [review]
See https://github.com/open-telemetry/opentelemetry-dotnet/pull/2949/files#r815222141

File: docs/trace/extending-the-sdk/MyResourceDetector.cs
Comment (Yun-Ting) [reply docs/trace/extending-the-sdk/MyResourceDetector.cs:21]: Added usage.

File: docs/trace/extending-the-sdk/MyResourceDetector.cs
Comment (Yun-Ting) [reply docs/trace/extending-the-sdk/MyResourceDetector.cs]: I've removed it.

File: docs/trace/extending-the-sdk/MyResourceDetector.cs
Comment (Yun-Ting) [reply docs/trace/extending-the-sdk/MyResourceDetector.cs]: I've done a minor refactor to the detect method.

File: src/OpenTelemetry/Resources/Resource.cs
Comment (cijothomas) [reply src/OpenTelemetry/Resources/Resource.cs:38]: After re-reading, and offline discussion, we'll have to expose this public, as Detectors need to return Resource, so need public ctor to do that.

File: docs/trace/extending-the-sdk/MyResourceDetector.cs
Comment (Yun-Ting) [reply docs/trace/extending-the-sdk/MyResourceDetector.cs:26]: Update the file to point to the example.

File: docs/trace/extending-the-sdk/MyResourceDetector.cs
Comment (cijothomas) [inline docs/trace/extending-the-sdk/MyResourceDetector.cs:28]: keep this as simple as possible. I'd recommend to simply return a new List of attributes. Don't need to show env variable parsing or anything here.

File: docs/trace/extending-the-sdk/MyResourceDectectorExtensions.cs
Comment (cijothomas) [inline docs/trace/extending-the-sdk/MyResourceDectectorExtensions.cs:19]: not really required, right? I'd skip this whole class.

File: docs/trace/extending-the-sdk/MyResourceDetector.cs
Comment (Yun-Ting) [reply docs/trace/extending-the-sdk/MyResourceDetector.cs:28]: I think it serves as a good example for the users to provide a more wholistic usage on where to get the resource attributes, at lease for me I think it is good to have information. But I'm open to other suggestions.

File: docs/trace/extending-the-sdk/MyResourceDectectorExtensions.cs
Comment (Yun-Ting) [reply docs/trace/extending-the-sdk/MyResourceDectectorExtensions.cs]: removed

File: docs/trace/extending-the-sdk/MyResourceDetector.cs
Comment (cijothomas) [reply docs/trace/extending-the-sdk/MyResourceDetector.cs:28]: eg: Sampler (doc has the simplest one, but links to other official ones from this repo)

**Review by cijothomas — APPROVED** [review]
LGTM.
Will wait to get one more approval before merge, as this is adding new public API very close to a stable release, and want to get one more review.

**Review by reyang — APPROVED** [review]
LGTM.

File: docs/trace/extending-the-sdk/README.md
Comment (reyang) [inline docs/trace/extending-the-sdk/README.md:7]:
```suggestion
* [Building your own resource detector](#resourcedetector)
```

File: docs/trace/extending-the-sdk/README.md
Comment (alanwest) [inline docs/trace/extending-the-sdk/README.md:299]:
```suggestion
## Resource Detector

OpenTelemetry .NET SDK provides a resource detector for detecting
resource information from the `OTEL_RESOURCE_ATTRIBUTES` and
`OTEL_SERVICE_NAME` environment variables.

Custom resource detectors can be implemented:
```

**Review by alanwest — APPROVED** [review]
LGTM

**Review by mic-max — APPROVED** [review]

**Review by cijothomas — APPROVED** [review]

**Pull request merged**
